### PR TITLE
Products & Product Variation: Allow alternative type for purchasable

### LINKED ISSUE #7510: Swift.DecodingError on purchasable products. Expected to decode Bool but found a Number.
Ref: p4a5px-2PC-p2

Some plugins seem to return `0` or  `1` rather than the expected  boolean for the [purchasable API response](https://woocommerce.github.io/woocommerce-rest-api-docs/#product-properties). This leads to problems loading content within the app as data can't be decoded:

```
let purchasable = try container.decode(Bool.self, forKey: .purchasable)
```

We can attempt and offer an alternative decoding if the expected one fails. For example, we offer alternative types for `salePrice` as is known that plugins alter this field:

```
        // Even though a plain install of WooCommerce Core provides string values,
        // some plugins alter the field value from String to Int or Decimal.
        let salePrice = container.failsafeDecodeIfPresent(targetType: String.self,
                                                          forKey: .salePrice,
                                                          shouldDecodeTargetTypeFirst: false,
                                                          alternativeTypes: [
                                                            .string(transform: { (onSale && $0.isEmpty) ? "0" : $0 }),
                                                            .decimal(transform: { NSDecimalNumber(decimal: $0).stringValue })])
            ?? ""
```

### PR DESCRIPTION
**Pull request by reginabally**
Closes: #7510 

## Description

Some plugins alter the field value from a boolean to an integer. This causes issues in loading products in the app. As suggested in #7510, we can attempt to offer an alternative decoding if the expected one fails.

### Changes:

* In `Networking/Product` - update decoding of purchasable to accept both Bool and an alternative integer value.
* In `Networking/ProductVariation` - update decoding of purchasable to accept both Bool and an alternative integer value.
* In `Networking.KeyedDecodingContainer+Woo` - add integer support to `alternativeTypes`

## Testing instructions

According to p4a5px-2PC-p2, we found out the [YITH WooCommerce Gift Cards](https://wordpress.org/plugins/yith-woocommerce-gift-cards/) plugin is returning `0` or `1` values for the `purchasable` field. This plugin will be used for this test.

1. Install the [YITH WooCommerce Gift Cards](https://wordpress.org/plugins/yith-woocommerce-gift-cards/) extension on your store
2. Create a gift card at WP Admin > YITH > Gift Cards. Enter an amount (e.g., 10) and any current balance (> 0).
3. Create a new product at WP Admin > Products > Add New.
4. Set the `Product data` to `Gift card` and add any Gift card amounts.
5. Check the product fields via WC REST API using the product ID (e.g., `../wp-json/wc/v3/products/212`). The `purchasable` field should return 1. (Note: if you set the current balance to 0 in step 2, `purchasable` will be 0)
6. In the app, open the Products tab
7. All products should be shown

## Screenshots

**Before**

**After**

---
- [X] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: Networking/Networking/Model/Product/Product.swift [hunk 1]
```diff
@@ -319,7 +319,12 @@ public struct Product: Codable, GeneratedCopiable, Equatable, GeneratedFakeable
                                                             .decimal(transform: { NSDecimalNumber(decimal: $0).stringValue })])
             ?? ""
 
-        let purchasable = try container.decode(Bool.self, forKey: .purchasable)
+        // Even though WooCommerce Core returns Bool values,
+        // some plugins aler the field value from Bool to Int (1 or 0)
```
Comment (iamgabrielma) [inline Networking/Networking/Model/Product/Product.swift]: Typo! 😛 

```suggestion
        // some plugins alter the field value from Bool to Int (1 or 0)
```

File: Networking/Networking/Model/Product/ProductVariation.swift [hunk 2]
```diff
@@ -194,7 +194,12 @@ public struct ProductVariation: Codable, GeneratedCopiable, Equatable, Generated
                                                             .decimal(transform: { NSDecimalNumber(decimal: $0).stringValue })])
             ?? ""
 
-        let purchasable = try container.decode(Bool.self, forKey: .purchasable)
+        // Even though WooCommerce Core returns Bool values,
+        // some plugins aler the field value from Bool to Int (1 or 0)
```
Comment (iamgabrielma) [inline Networking/Networking/Model/Product/ProductVariation.swift]: Same here: 

```suggestion
        // some plugins alter the field value from Bool to Int (1 or 0)
```

**Review by iamgabrielma — CHANGES_REQUESTED** [review]
Thanks for contributing @reginabally ! The changes work as expected and the product list loads despite having a different decoding type than expected by core.

I have 2 suggestions!

> The favorable alternative type is an integer, but I couldn't find a way to add it to the list of alternativeTypes because Networking.KeyedDecodingContainer+Woo doesn't support integers now.

1. I agree that the best type to check would be Integer here, as they're using either 0 or 1 for the boolean type. You're also right that the case for `.integer` type is not supported now, so, should we add it? AFAIK you could add the extra case to the `AlternativeDecodingType<T>` enum, so we could do something like: 

```
case integer(transform: (Int) -> T)
```
Then you can follow the rest of the pattern for `failsafeDecodeIfPresent<T>` when the compiler complains that switch cases must be exhaustive.

Once done, we can use the new alternative type on the `.purchasable` decoding, for example (this needs to be tested 😝 ):

```
.integer(transform: { ($0 as NSNumber).boolValue })
```
2. We could add some Unit Tests here to be sure that decodes the data as expected for both expected values (Bool) and non-expected values (Int), as well to assure we're not breaking something else, and decoding continues to work in the future if we keep changing things. You can find some examples on how to add these in the repo, like [here](https://github.com/woocommerce/woocommerce-ios/pull/7109) or [here](https://github.com/woocommerce/woocommerce-ios/pull/5898)

Let me know if you have any question!

File: Networking/Networking/Model/Product/Product.swift (hunk 1 above)
Comment (reginabally) [reply Networking/Networking/Model/Product/Product.swift]: Thanks for spotting this! Typo corrected.

File: Networking/Networking/Model/Product/ProductVariation.swift (hunk 2 above)
Comment (reginabally) [reply Networking/Networking/Model/Product/ProductVariation.swift]: Typo corrected.

**Comment by reginabally** [comment]
Thanks for reviewing my changes, @iamgabrielma!

I added integer support to `Networking.KeyedDecodingContainer+Woo`, and added the `purchasable` field in the Unit Test `ProductMapperTests`.

The unit test result was okay:

```
est Suite 'Selected tests' started at 2023-01-20 15:55:57.793
Test Suite 'NetworkingTests.xctest' started at 2023-01-20 15:55:57.794
Test Suite 'ProductMapperTests' started at 2023-01-20 15:55:57.794
Test Case '-[NetworkingTests.ProductMapperTests test_that_product_alternative_types_are_properly_parsed]' started.
2023-01-20 15:55:57.824877+0800 xctest[30982:341115] Message: Tag aztec.htmltag.rootnode invalid

Test Case '-[NetworkingTests.ProductMapperTests test_that_product_alternative_types_are_properly_parsed]' passed (0.040 seconds).
Test Suite 'ProductMapperTests' passed at 2023-01-20 15:55:57.835.
	 Executed 1 test, with 0 failures (0 unexpected) in 0.040 (0.041) seconds
Test Suite 'NetworkingTests.xctest' passed at 2023-01-20 15:55:57.835.
	 Executed 1 test, with 0 failures (0 unexpected) in 0.040 (0.041) seconds
Test Suite 'Selected tests' passed at 2023-01-20 15:55:57.835.
	 Executed 1 test, with 0 failures (0 unexpected) in 0.040 (0.042) seconds
Program ended with exit code: 0
```

When I have `purchasable` set to 1 in `product-alternative-types.json`:

```
"purchasable": 1,
```

**Comment by reginabally** [comment]
I am closing this for #8718.

**Pull request closed without merging**